driver: uart: stm32: Add workaround for DMAT errata in low-power modes

ABESTM opened the pull request:
## Problem
UART driver for STM32U5 implements workaround for following errata "USART does not generate DMA requests after setting/clearing DMAT bit". It uses first approach described in errata, which is not clearing the DMAT bit.
This causes the DMA TX request to be pending to DMA controller and basically prevents entering STOP low-power mode. In other words, it keeps internal bus clock and increases power consumption.
## Solution
The workaround for low-power mode is to send first byte by polling and configure the DMA TX buffer to start at second byte of the buffer.
## Open points
1.  I think it makes sense to only enable this workaround when low-power modes are used (`CONFIG_PM=y`). I'm not sure if this should be configured automatically, or if user should enable this explicitly through KConfig. This should also automatically disable the `CONFIG_UART_STM32U5_ERRATA_DMAT`.
2. For workaround to work, we need to minimize the delay between sending the first byte and starting the DMA. The DMA needs to be started before the first byte is sent. Currently this is done by disabling interrupts globally. Maybe there could be better solution?

ABESTM committed:
driver: uart: stm32: Add workaround for DMAT errata in low-power modes

This change adds additional workaround for following errata:
"USART does not generate DMA requests after setting/clearing DMAT bit"
Instead of keepint DMAT bit set, it sends first byte by polling
in firmware. This prevents additional power consumption in STOP mode,
caused by keeping DMAT bit set.

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

FRASTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -1651,34 +1651,44 @@ static int uart_stm32_async_tx(const struct device *dev,
+	key = irq_lock();
+
+	LL_USART_TransmitData8(usart, tx_data[0]);
+
```
After transmitting tx_data[0],    this data should not be transmitted again by the DMA

ABESTM replied:
The `data->dma_tx.blk_cfg.source_address` and `data->dma_tx.blk_cfg.block_size` should be modified accordingly, so the first byte is not transmitted by DMA. These variables should be used for DMA configuration, if I'm correct.
I kept `data->dma_tx.buffer` and `data->dma_tx.buffer_length` the same way, including the first byte.

ABESTM commented:
I modified the KConfig to implement this workaround only when `CONFIG_PM=y`. I switched the `UART_STM32U5_ERRATA_DMAT` to a choice type. There are 3 options now:
1. `UART_STM32U5_ERRATA_DMAT_NONE` - no workaround, default for unaffected devices
2. `UART_STM32U5_ERRATA_DMAT_NOCLEAR` - keeping DMAT bit set, default for affected devices without `CONFIG_PM=y`. This was `UART_STM32U5_ERRATA_DMAT=y` previously
3. `UART_STM32U5_ERRATA_DMAT_LOWPOWER` - new workaround sending first byte by polling

This also allows user to overwrite default configuration if needed.

erwango reviewed (commented):
Minor non blocking comments, but otherwise LGTM if @FRASTM approve

ABESTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -1650,34 +1651,58 @@ static int uart_stm32_async_tx(const struct device *dev,
+		if (ret != 0) {
+			LOG_ERR("dma tx config error!");
+			return -EINVAL;
+		}
 
-	if (dma_start(data->dma_tx.dma_dev, data->dma_tx.dma_channel)) {
-		LOG_ERR("UART err: TX DMA start failed!");
-		return -EFAULT;
-	}
+		if (dma_start(data->dma_tx.dma_dev, data->dma_tx.dma_channel)) {
+			LOG_ERR("UART err: TX DMA start failed!");
+			return -EFAULT;
+		}
 
-	/* Start TX timer */
-	async_timer_start(&data->dma_tx.timeout_work, data->dma_tx.timeout);
+		/* Start TX timer */
+		async_timer_start(&data->dma_tx.timeout_work, data->dma_tx.timeout);
```
Maybe I could rewrite like this, and it would make it more readable and also the indentation wouldn't be so messy?
```c
#if CONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER
	if (data->dma_tx.buffer_length > 1) {
...
```
to
```c
if(!IS_ENABLED(CONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER) || (data->dma_tx.buffer_length > 1)){
```

mathieuchopstm commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,15 +25,41 @@ config UART_STM32
+	prompt "No workaround"
+	help
+		No workaround selected or SoC not affected.
+
+config UART_STM32U5_ERRATA_DMAT_LOWPOWER
+	bool
+	prompt "Sending first byte by polling"
+	depends on SOC_STM32U575XX || SOC_STM32U585XX || \
+		   SOC_STM32H562XX || SOC_STM32H563XX || SOC_STM32H573XX
+	help
+	  This option sends first byte via software polling and
+	  rest of the buffer is sent via DMA. This option is suitable
+	  for STOP low-power modes.
+
+config UART_STM32U5_ERRATA_DMAT_NOCLEAR
+	bool
+	prompt "Not clearing DMAT"
+	depends on SOC_STM32U575XX || SOC_STM32U585XX || \
+		   SOC_STM32H562XX || SOC_STM32H563XX || SOC_STM32H573XX
+	help
+	  This option keeps DMAT bit set. This may cause additional power
+	  consumption in STOP low-power modes.
```
I think the dependency check can be expressed as follows to avoid duplication:
```suggestion
if SOC_STM32U575XX || SOC_STM32U585XX || \
   SOC_STM32H562XX || SOC_STM32H563XX || SOC_STM32H573XX

config UART_STM32U5_ERRATA_DMAT_LOWPOWER
	bool
	prompt "Sending first byte by polling"
	help
	  This option sends first byte via software polling and
	  rest of the buffer is sent via DMA. This option is suitable
	  for STOP low-power modes.

config UART_STM32U5_ERRATA_DMAT_NOCLEAR
	bool
	prompt "Not clearing DMAT"
	help
	  This option keeps DMAT bit set. This may cause additional power
	  consumption in STOP low-power modes.

endif # U575 || U585 || H562 || H563 || H573
```

mathieuchopstm commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,15 +25,41 @@ config UART_STM32
+		No workaround selected or SoC not affected.
+
+config UART_STM32U5_ERRATA_DMAT_LOWPOWER
+	bool
+	prompt "Sending first byte by polling"
```
\+ wording
```suggestion
	bool "Send first byte by polling"
```

mathieuchopstm commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,15 +25,41 @@ config UART_STM32
+	  for STOP low-power modes.
+
+config UART_STM32U5_ERRATA_DMAT_NOCLEAR
+	bool
+	prompt "Not clearing DMAT"
```
\+ wording
```suggestion
	bool
	prompt "Do not clear DMAT"
```
or:
```suggestion
	bool
	prompt "Never clear DMAT"
```

mathieuchopstm commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,15 +25,41 @@ config UART_STM32
-	depends on SOC_STM32U575XX || SOC_STM32U585XX || \
-		   SOC_STM32H562XX || SOC_STM32H563XX || SOC_STM32H573XX
+choice UART_STM32U5_ERRATA_DMAT
+	prompt "DMAT workaround for selected devices"
```
Proposal:
```suggestion
	prompt "Workaround for DMAT errata on affected devices"
```

mathieuchopstm reviewed (commented):
Otherwise LGTM

ABESTM commented:
Thank you for the feedback. I updated the PR.

mathieuchopstm commented on drivers/serial/uart_stm32.c:
```diff
@@ -1650,34 +1649,56 @@ static int uart_stm32_async_tx(const struct device *dev,
-	/* set source address */
-	data->dma_tx.blk_cfg.source_address = (uint32_t)data->dma_tx.buffer;
-	data->dma_tx.blk_cfg.block_size = data->dma_tx.buffer_length;
+	if (!IS_ENABLED(CONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER) ||
```
Nit: small comment to explain this condition
```suggestion
        /**
         * Setup DMA descriptor for TX.
         * If DMAT low-power errata workaround is enabled,
         * we send the first byte using polling and the rest
         * using DMA; as such, single-byte transfers use only
         * polling and don't need to prepare a DMA descriptor.
         * In other configurations, the DMA is always used.
         */
	if (!IS_ENABLED(CONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER) ||
```

mathieuchopstm commented on drivers/serial/uart_stm32.c:
```diff
@@ -1532,8 +1532,6 @@ void uart_stm32_dma_tx_cb(const struct device *dma_dev, void *user_data,
 	}
 
-	data->dma_tx.buffer_length = 0;
-
```
I'm probably missing some context, but I don't see what makes this change required - can you elaborate a little?

ABESTM replied:
I will probably move this to separate commit to clarify as well in commit message.
The `buffer_length != 0` is used to check if there is transfer ongoing inside the `uart_stm32_async_tx`. This callback clears it to 0 when TX DMA is finished, but before the USART send all the data from shift register.
The problem is that there is another callback `async_evt_tx_done`, which is called when USART finished sending all the bytes and this clears `buffer_length` as well.
If the `uart_stm32_async_tx` would be called after `uart_stm32_async_tx`, but before `async_evt_tx_done` there could be race condition on `buffer_length`.

mathieuchopstm reviewed (commented):
One minor nit + one question but otherwise LGTM.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1650,34 +1649,56 @@ static int uart_stm32_async_tx(const struct device *dev,
-	data->dma_tx.blk_cfg.block_size = data->dma_tx.buffer_length;
+	if (!IS_ENABLED(CONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER) ||
+	    data->dma_tx.buffer_length > 1) {
+#if CONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER
+		/* set source address */
+		data->dma_tx.blk_cfg.source_address = ((uint32_t)data->dma_tx.buffer) + 1;
+		data->dma_tx.blk_cfg.block_size = data->dma_tx.buffer_length - 1;
+#else
+		/* set source address */
+		data->dma_tx.blk_cfg.source_address = ((uint32_t)data->dma_tx.buffer);
+		data->dma_tx.blk_cfg.block_size = data->dma_tx.buffer_length;
+#endif
```
Non blocking: 
```suggestion
		if (IS_ENABLED(CONFIG_UART_STM32U5_ERRATA_DMAT_LOWPOWER) {
				/* set source address */
				data->dma_tx.blk_cfg.source_address = ((uint32_t)data->dma_tx.buffer) + 1;
				data->dma_tx.blk_cfg.block_size = data->dma_tx.buffer_length - 1;
		} else {
				/* set source address */
				data->dma_tx.blk_cfg.source_address = ((uint32_t)data->dma_tx.buffer);
				data->dma_tx.blk_cfg.block_size = data->dma_tx.buffer_length;
		}
```

erwango commented on drivers/serial/Kconfig.stm32:
```diff
@@ -25,15 +25,39 @@ config UART_STM32
 	          Errata Sheet 0565 § 2.14.1 and §2.15.1 for stm32h56x/h57x
 
-endif
+config UART_STM32U5_ERRATA_DMAT_NONE
+	bool "No workaround"
+	help
+	  No workaround selected or SoC not affected.
```
Now that the choice is made specific to affected SoC, this specific symbol is likely not useful anymore. Did you had a try removing it and compile on non-affected SoC ?

ABESTM committed:
driver: uart: stm32: Clear buffer length only in UART TC event

The buffer_length != 0 is used to check if there is transfer ongoing
inside the uart_stm32_async_tx function. uart_stm32_dma_tx_cb clears
it to 0 when TX DMA is finished, but before the USART send all
the data from shift register.
buffer_length is also cleared in callback async_evt_tx_done,
which is called when USART finished sending all the bytes.
If the uart_stm32_async_tx would be called after
uart_stm32_async_tx, but before async_evt_tx_done there
could be race condition on buffer_length.

Signed-off-by: Adam BERLINGER <adam.berlinger@st.com>
